File: Pharmaceutical_Company_Example/03_Behavioral/R/Visitor.R
# Define Element Interface for Regulatory Steps
RegulatoryStep <- setRefClass(
  "RegulatoryStep",
  methods = list(
    accept = function(visitor) {
      stop("This method should be overridden by concrete subclasses!")
    }
  )
)

# Concrete Elements
DocumentationSubmission <- setRefClass(
  "DocumentationSubmission",
  contains = "RegulatoryStep",
  methods = list(
    accept = function(visitor) {
      visitor$visit_documentation_submission(.self)
    }
  )
)

ClinicalTrialReview <- setRefClass(
  "ClinicalTrialReview",
  contains = "RegulatoryStep",
  methods = list(
    accept = function(visitor) {
      visitor$visit_clinical_trial_review(.self)
    }
  )
)

SafetyEvaluation <- setRefClass(
  "SafetyEvaluation",
  contains = "RegulatoryStep",
  methods = list(
    accept = function(visitor) {
      visitor$visit_safety_evaluation(.self)
    }
  )
)

# Visitor Interface
Visitor <- setRefClass(
  "Visitor",
  methods = list(
    visit_documentation_submission = function(documentation) {
      stop("This method should be overridden by concrete visitors!")
    },
    visit_clinical_trial_review = function(trial_review) {
      stop("This method should be overridden by concrete visitors!")
    },
    visit_safety_evaluation = function(safety_evaluation) {
      stop("This method should be overridden by concrete visitors!")
    }
  )
)

# Concrete Visitor Auditor
Auditor <- setRefClass(
  "Auditor",
  contains = "Visitor",
  methods = list(
    visit_documentation_submission = function(documentation) {
      message("Auditing documentation submission...")
    },
    visit_clinical_trial_review = function(trial_review) {
      message("Auditing clinical trial review process...")
    },
    visit_safety_evaluation = function(safety_evaluation) {
      message("Auditing safety evaluation...")
    }
  )
)

# Concrete Visitor Report Generator
ReportGenerator <- setRefClass(
  "ReportGenerator",
  contains = "Visitor",
  methods = list(
    visit_documentation_submission = function(documentation) {
      message("Generating report for documentation submission...")
    },
    visit_clinical_trial_review = function(trial_review) {
      message("Generating report for clinical trial review...")
    },
    visit_safety_evaluation = function(safety_evaluation) {
      message("Generating report for safety evaluation...")
    }
  )
)

# Concrete Visitor Approval Checker
ApprovalChecker <- setRefClass(
  "ApprovalChecker",
  contains = "Visitor",
  methods = list(
    visit_documentation_submission = function(documentation) {
      message("Checking approval status of documentation submission...")
    },
    visit_clinical_trial_review = function(trial_review) {
      message("Checking approval status of clinical trial review...")
    },
    visit_safety_evaluation = function(safety_evaluation) {
      message("Checking approval status of safety evaluation...")
    }
  )
)

# Define the regulatory steps
documentation <- DocumentationSubmission$new()
clinical_trials <- ClinicalTrialReview$new()
safety_eval <- SafetyEvaluation$new()

# Define visitors
auditor <- Auditor$new()
report_generator <- ReportGenerator$new()
approval_checker <- ApprovalChecker$new()

# Use the Auditor visitor on different steps
documentation$accept(auditor)
clinical_trials$accept(auditor)
safety_eval$accept(auditor)

# Use the Report Generator visitor on different steps
documentation$accept(report_generator)
clinical_trials$accept(report_generator)
safety_eval$accept(report_generator)

# Use the Approval Checker visitor on different steps
documentation$accept(approval_checker)
clinical_trials$accept(approval_checker)
safety_eval$accept(approval_checker)

File: Pharmaceutical_Company_Example/03_Behavioral/JavaScript/Visitor.js
// Define Element Interface for Regulatory Steps
class RegulatoryStep {
  accept(visitor) {
    throw new Error("This method should be overridden by concrete classes!");
  }
}

// Concrete Elements
class DocumentationSubmission extends RegulatoryStep {
  accept(visitor) {
    visitor.visitDocumentationSubmission(this);
  }
}

class ClinicalTrialReview extends RegulatoryStep {
  accept(visitor) {
    visitor.visitClinicalTrialReview(this);
  }
}

class SafetyEvaluation extends RegulatoryStep {
  accept(visitor) {
    visitor.visitSafetyEvaluation(this);
  }
}

// Visitor Interface
class Visitor {
  visitDocumentationSubmission(documentation) {
    throw new Error("This method should be overridden by concrete classes!");
  }

  visitClinicalTrialReview(trialReview) {
    throw new Error("This method should be overridden by concrete classes!");
  }

  visitSafetyEvaluation(safetyEvaluation) {
    throw new Error("This method should be overridden by concrete classes!");
  }
}

// Concrete Visitor Auditor
class Auditor extends Visitor {
  visitDocumentationSubmission(documentation) {
    console.log("Auditing documentation submission...");
  }

  visitClinicalTrialReview(trialReview) {
    console.log("Auditing clinical trial review process...");
  }

  visitSafetyEvaluation(safetyEvaluation) {
    console.log("Auditing safety evaluation...");
  }
}

// Concrete Visitor Report Generator
class ReportGenerator extends Visitor {
  visitDocumentationSubmission(documentation) {
    console.log("Generating report for documentation submission...");
  }

  visitClinicalTrialReview(trialReview) {
    console.log("Generating report for clinical trial review...");
  }

  visitSafetyEvaluation(safetyEvaluation) {
    console.log("Generating report for safety evaluation...");
  }
}

// Concrete Visitor Approval Checker
class ApprovalChecker extends Visitor {
  visitDocumentationSubmission(documentation) {
    console.log("Checking approval status of documentation submission...");
  }

  visitClinicalTrialReview(trialReview) {
    console.log("Checking approval status of clinical trial review...");
  }

  visitSafetyEvaluation(safetyEvaluation) {
    console.log("Checking approval status of safety evaluation...");
  }
}

// Define the regulatory steps
const documentation = new DocumentationSubmission();
const clinicalTrials = new ClinicalTrialReview();
const safetyEval = new SafetyEvaluation();

// Define visitors
const auditor = new Auditor();
const reportGenerator = new ReportGenerator();
const approvalChecker = new ApprovalChecker();

// Use the Auditor visitor on different steps
documentation.accept(auditor);
clinicalTrials.accept(auditor);
safetyEval.accept(auditor);

// Use the Report Generator visitor on different steps
documentation.accept(reportGenerator);
clinicalTrials.accept(reportGenerator);
safetyEval.accept(reportGenerator);

// Use the Approval Checker visitor on different steps
documentation.accept(approvalChecker);
clinicalTrials.accept(approvalChecker);
safetyEval.accept(approvalChecker);
